### PR TITLE
Feature/asynchronous authorization rules

### PR DESCRIPTION
**Pull request by luizfbicalho**
I just added the basic idea so we could talk about the whole spec. 
Didn't add the tests because of that

#2801

### CODE REVIEW

**Comment by luizfbicalho** [comment]
> @luizbicalhoagl, I don't have any immediate concerns about the changes, as long as everything still works 😄
> 
> And by "works" what I mean is that pages that use the API synchronously today keep working, as do any consumers using the API in an async manner.

Thanks for the feedback 

I did my best to create the async methods without impacting on the previous ones

If it's ok for you I can create the tests for some async authorization rules.
Do you have any scenario in mind that needs attention?

@StefanOssendorf do you have something to add?

**Comment by rockfordlhotka** [comment]
The original request came from @mtavares628, so perhaps he has a good example for testing?

I think the real scenario is an authn rule that needs to do some async data portal command - like read from a database or other backend service. You can simulate this by creating a rule that does a `FetchAsync` on the data portal - perhaps on the logical server-side `Fetch` operation do an `await Task.Delay(10)` or something.

**Comment by mtavares628** [comment]
@luizfbicalho, @rockfordlhotka hit the nail on the head with my use case.  I want to create an authorization rule that is able to perform a FetchAsync or ExecuteAsync to retrieve a value back from the database to determine whether the user is authorized or not. The criteria for that fetch would be passed in to the HasPermissionAsync method as the criteria object.  You could probably extend the tests from [PR #1226](https://github.com/MarimerLLC/csla/pull/1226/files) to simulate the fetch. While it just checks the type of the criteria object in those tests, you could instead use the value from the criteria to perform the fetch.

**Comment by rockfordlhotka** [comment]
@luizfbicalho let me know when you have added tests for this

**Comment by luizfbicalho** [comment]
> @luizfbicalho let me know when you have added tests for this

I'll finish them thiis week

**Comment by luizfbicalho** [comment]
> @luizfbicalho let me know when you have added tests for this


I added the tests in a way that I'm validating that sync rules can be called async and worked the same way, and async rules can't be called sync, throwing exception

File: Source/Csla/Rules/IAuthorizationRuleBase.cs [hunk 1]
```diff
@@ -36,4 +31,31 @@ public interface IAuthorizationRule
     /// </summary>
     bool CacheResult { get; }
   }
+
+  /// <summary>
+  /// Interface defining an authorization
+  /// rule implementation.
+  /// </summary>
+  public interface IAuthorizationRule : IAuthorizationRuleBase
```
Comment (rockfordlhotka) [inline Source/Csla/Rules/IAuthorizationRuleBase.cs:39]: This interface should be in its own file

File: Source/Csla/Rules/IAuthorizationRuleBase.cs [hunk 2]
```diff
@@ -36,4 +31,31 @@ public interface IAuthorizationRule
...
+    /// </summary>
+    /// <param name="context">Rule context object.</param>
+    void Execute(IAuthorizationContext context);
+  }
+
+  /// <summary>
+  /// Interface defining an authorization
+  /// rule implementation.
+  /// </summary>
+  public interface IAuthorizationRuleAsync : IAuthorizationRuleBase
```
Comment (rockfordlhotka) [inline Source/Csla/Rules/IAuthorizationRuleBase.cs:52]: This interface also should be in its own file

**Review by rockfordlhotka — CHANGES_REQUESTED** [review]
We need to add an `AuthorizationRuleAsync` class to provide basic implementation of the `IAuthorizationRuleAsync` interface - like the existing `AuthorizationRule` class.

**Review by rockfordlhotka — APPROVED** [review]

**Pull request merged**